SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestRemotePossession.h"
+
+#include "Kismet/GameplayStatics.h"
+#include "TestPossessionPawn.h"
+#include "TestPossessionPlayerController.h"
+
+const float ASpatialTestRemotePossession::MaxWaitTime = 2.0f;
+
+ASpatialTestRemotePossession::ASpatialTestRemotePossession()
+	: Super()
+	, WaitTime(0.0f)
+{
+	Author = "Jay";
+	Description = TEXT("Test Actor Remote Possession");
+}
+
+ATestPossessionPawn* ASpatialTestRemotePossession::GetPawn()
+{
+	return Cast<ATestPossessionPawn>(UGameplayStatics::GetActorOfClass(GetWorld(), ATestPossessionPawn::StaticClass()));
+}
+
+void ASpatialTestRemotePossession::AddWaitStep(const FWorkerDefinition& Worker)
```

Comment:
This kind of step, while convenient for testing, should be avoided IMO.
Our functional test should also serve as a reference for how a feature is to be used.
I hope we won't ask our users to add arbitrary waiting in their code.

Reply:
Sorry, can you tell me more about IMO?
From my understanding, remotePossess is not immediate. Migration to the target worker should have a roundtrip latency.
If not use WaitStep, how can we check it possessed or not? 
The flow controllers are not on the same worker-side, how to synchronize the time process?

Reply:
Recommend run lock policy case manually instead of automatic, we only check if the Actor locked and return with a log like:
Actor TestPossessionPlayerController_0 (9013) cannot migration because of locked [C:\io\UE4\4.25\Engine\Plugins\UnrealGDK\SpatialGDK\Source\SpatialGDK\Private\Utils\SpatialLoadBalancingHandler.cpp(61)]
Have no idea how to check from the log.

Reply:
IMO == In My Opinion

You can provide 3 callbacks to AddStep.
The first one is returns a bool. It will be called until it returns true. That is where you can do the waiting, you can check there that something you are expecting happens. Only the server where the pawn to be possessed is authoritative would do that check, meaning other servers will instantly complete the step, but they will wait for all the servers to have completed the step before moving forward.

For the lock step, using a timeout is ok. If the failed operation is checked as successful in another test, checking that something does not happen in different conditions through a timeout is Ok.
Unreal has something for expected errors but I'm afraid that it's not in FunctionalTest, see FAutomationTestBase::AddExpectedError

Reply:
Thanks for FAutomationTestBase::AddExpectedError. But I still have no idea about using it for this lock policy case.
I change some check results steps to IsReady callback.
But I still use `WaitStep` to prepare the Possess, because I have no idea how to check the Controller is in the right worker.